### PR TITLE
Broker discovery

### PR DESCRIPTION
**Pull request by jwilder**
This sends data node urls via the broker heartbeat from each data node.  The urls are tracked on the broker to support simpler cluster setup as well as distributed queries.

This is a prerequisite for splitting data and broker nodes.

cc @benbjohnson @otoolep

### CODE REVIEW

File: cmd/influxd/run.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 	"github.com/influxdb/influxdb/udp"
 )
 
-func Run(config *Config, join, version string) (*messaging.Broker, *influxdb.Server) {
+func Run(config *Config, join, version string) (*messaging.Broker, *influxdb.Server, *raft.Log) {
```
Comment (otoolep) [inline cmd/influxd/run.go]: I feel like the return values are getting a little bit out of control. Perhaps we need a more generic `node` type which returned by `Run()`. This `node` type would have a `Close()` method, which would take of closing whatever was opened. Sometimes `node` would have a broker, sometimes a log, sometimes a server, sometimes all, etc.

Comment (jwilder) [inline cmd/influxd/run.go]: Agreed.  We started down that path in https://github.com/influxdb/influxdb/pull/2118.  We're moving in that direction but planning to do that in a separate PR.

File: messaging/broker.go [hunk 2]
```diff
@@ -619,9 +631,13 @@ const DefaultMaxSegmentSize = 10 * 1024 * 1024 // 10MB
 type Topic struct {
 	mu    sync.Mutex
 	id    uint64 // unique identifier
-	index uint64 // highest index replicated
```
Comment (otoolep) [inline messaging/broker.go]: Was this comment incorrect, or does `index` now mean something new?

File: messaging/broker_test.go [hunk 3]
```diff
@@ -37,7 +42,7 @@ func TestBroker_Close_ErrClosed(t *testing.T) {
 	b := NewBroker()
 	b.Close()
 	if err := b.Broker.Close(); err != messaging.ErrClosed {
-		t.Fatalf("unexpected error: %s", err)
```
Comment (otoolep) [inline messaging/broker_test.go]: Looks like a rogue ` in there.

File: messaging/broker_test.go [hunk 4]
```diff
@@ -37,7 +42,7 @@ func TestBroker_Close_ErrClosed(t *testing.T) {
...
+		t.Fatalf("unexpected error: %s`", err)
```
Comment (otoolep) [inline messaging/broker_test.go]: Looks like a rogue ` in there.

File: messaging/broker_test.go [hunk 5]
```diff
@@ -124,13 +129,19 @@ func TestBroker_Apply_SetMaxTopicIndex(t *testing.T) {
 	if err := b.Apply(&messaging.Message{
 		Index: 2,
 		Type:  messaging.SetTopicMaxIndexMessageType,
-		Data:  []byte{0, 0, 0, 0, 0, 0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 5}, // topicID=20, index=5
+		Data: []byte{0, 0, 0, 0, 0, 0, 0, 20, 0, 0, 0, 0, 0, 0, 0, 5, // topicID=20, index=5,
+			0, 26, 104, 116, 116, 112, 58, 47, 47, 108, 111, 99, 97, 108, //len=26, url="http://localhost:1234/data"
```
Comment (otoolep) [inline messaging/broker_test.go]: I'll take your word for it. :-)

We could have made this slice, or at least part of it, but appending `[]byte("http://")`.

File: messaging/client.go [hunk 6]
```diff
@@ -335,12 +339,12 @@ func (c *Client) do(method, path string, values url.Values, contentType string,
 }
 
 // Conn returns a connection to the broker for a given topic.
-func (c *Client) Conn(topicID uint64) *Conn {
+func (c *Client) Conn(topicID uint64, dataURL *url.URL) *Conn {
```
Comment (otoolep) [inline messaging/client.go]: Shouldn't this function be named `NewConn`?

Comment (otoolep) [inline messaging/client.go]: I know it's a method, not a free function, but I still think the name could be different.

File: messaging/client.go [hunk 7]
```diff
@@ -531,6 +537,20 @@ func (c *Conn) close() error {
...
 
+// heartbeater periodically heartbeats the broker it's index
+func (c *Conn) heartbeater(closing chan struct{}) {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-closing:
+			return
+		case <-time.After(DefaultHeartbeatInterval):
```
Comment (otoolep) [inline messaging/client.go]: We tend to use `ticker` here. It gives a better guarantee of timing I believe.

File: messaging/broker.go (hunk 2 above)
Comment (benbjohnson) [inline messaging/broker.go]: The field was accidentally being overloaded before and it stored the last index written to the topic at first and then used for heartbeats. The heartbeat tracking was moved to the `indexByURL` field.

File: messaging/client.go (hunk 6 above)
Comment (otoolep) [inline messaging/client.go]: Can `dataURL` be different across calls to `Conn`? If not, why is it a parameter in the function call? Should it be set when the `Client` object is created?

File: server.go [hunk 8]
```diff
@@ -980,7 +980,7 @@ func (s *Server) ShardGroups(database string) ([]*ShardGroup, error) {
 
 // CreateShardGroupIfNotExists creates the shard group for a retention policy for the interval a timestamp falls into.
 func (s *Server) CreateShardGroupIfNotExists(database, policy string, timestamp time.Time) error {
-	c := &createShardGroupIfNotExistsCommand{Database: database, Policy: policy, Timestamp: timestamp}
+	c := &createShardGroupIfNotExistsCommand{Database: database, Policy: policy, Timestamp: timestamp, URL: s.URL()}
```
Comment (otoolep) [inline server.go]: Why does this call need to thread the URL through it? When the "create shard" group command is processed, the server knows its own URL. It doesn't need it in the command, right?

**Comment by otoolep** [comment]
Does our design allow a node's URL to change? If so, using a node's URL as a key in a map could be problematic. If it does not allow a node's URL to change, then why continually send it over the heartbeat channel?

File: server.go (hunk 8 above)
Comment (jwilder) [inline server.go]: This was to avoid a deadlock in the server when processing this command.  The issue is that s.URL() locks the mutex that is already locked here: https://github.com/influxdb/influxdb/blob/master/server.go#L3190

Comment (otoolep) [inline server.go]: OK.

Is the URL of the server mutable? This is a big question, since if it is not, locking doesn't matter after the initial set.

Comment (jwilder) [inline server.go]: Looking at this again, I believe I can rework it to avoid the deadlock w/o adding it to the `CreateShardGroupsIfNotExists` command.  Will fix.

File: messaging/client.go (hunk 6 above)
Comment (jwilder) [inline messaging/client.go]: Good point.  Will look into moving it where the `Client` is created.

**Comment by jwilder** [comment]
@otoolep Updated based on your feedback.

File: messaging/client.go [hunk 9]
```diff
@@ -20,19 +20,24 @@ import (
...
 type Client struct {
-	mu    sync.Mutex
-	path  string    // config file path
-	conns []*Conn   // all connections opened by client
-	url   url.URL   // current known leader URL
-	urls  []url.URL // list of available broker URLs
+	mu      sync.Mutex
+	path    string    // config file path
+	conns   []*Conn   // all connections opened by client
+	url     url.URL   // current known leader URL
```
Comment (otoolep) [inline messaging/client.go]: This was there before, but we need to come back and fix this naming. `url` and `urls` are not great names on the same object. As names they don't mean much, and are too similar. They need to become `leaderURL` and `brokerURLs`. 

Not something required for this PR, one of us can fix it up afterwards.

File: messaging/client.go [hunk 10]
```diff
@@ -531,6 +544,21 @@ func (c *Conn) close() error {
 	return nil
 }
 
+// heartbeater periodically heartbeats the broker it's index
```
Comment (otoolep) [inline messaging/client.go]: Nit: it's -> its.

**Comment by otoolep** [comment]
OK, makes sense to me. Just some nits, which I don't need to re-review. You may wish to get this blessed by @benbjohnson, whatever you think.

Don't forget the CHANGELOG -- you could probably call it a "feature".

File: cmd/influxd/restore_test.go [hunk 11]
```diff
@@ -84,6 +88,7 @@ func TestRestoreCommand(t *testing.T) {
 	f.Close()
 
 	// Stop server.
+	l.Close()
 	s.Close()
 	b.Close()
```
Comment (benbjohnson) [inline cmd/influxd/restore_test.go]: Can you defer the closing right after you validate that `b`, `s`, & `l` are non-nil?

File: cmd/influxd/restore_test.go [hunk 12]
```diff
@@ -135,6 +140,11 @@ func TestRestoreCommand(t *testing.T) {
 	} else if !reflect.DeepEqual(v, map[string]interface{}{"value": float64(1000)}) {
 		t.Fatalf("read series(1) mismatch: %#v", v)
 	}
+
+	// Stop server.
+	l.Close()
+	s.Close()
+	b.Close()
```
Comment (benbjohnson) [inline cmd/influxd/restore_test.go]: Defer these as well.

File: cmd/influxd/run.go [hunk 13]
```diff
@@ -327,7 +327,7 @@ func openServer(config *Config, b *influxdb.Broker, initServer, initBroker bool,
 	}
 
 	// Create messaging client to the brokers.
-	c := influxdb.NewMessagingClient()
+	c := influxdb.NewMessagingClient(config.DataURL())
```
Comment (benbjohnson) [inline cmd/influxd/run.go]: Shouldn't this be `clientJoinURLs`?

File: cmd/influxd/server_integration_test.go [hunk 14]
```diff
@@ -66,6 +68,26 @@ type Node struct {
 // Cluster represents a multi-node cluster.
 type Cluster []*Node
```
Comment (benbjohnson) [inline cmd/influxd/server_integration_test.go]: Can you change the `Cluster` to be:

```go
type Cluster struct {
    Nodes []*Node
}
```

The `range *c` is confusing below and I think it would make more sense to have `range c.Nodes`.

File: messaging/broker_test.go [hunk 15]
```diff
@@ -19,9 +19,14 @@ import (
 	"github.com/influxdb/influxdb/raft"
 )
 
+var (
+	testDataURL *url.URL
+)
```
Comment (benbjohnson) [inline messaging/broker_test.go]: Can you move remove the block since it's only one `var`.

File: cmd/influxd/run.go (hunk 13 above)
Comment (benbjohnson) [inline cmd/influxd/run.go]: n/m, i see the function signature below.

File: messaging/client.go [hunk 16]
```diff
@@ -531,6 +544,21 @@ func (c *Conn) close() error {
 	return nil
 }
 
+// heartbeater periodically heartbeats the broker its index
+func (c *Conn) heartbeater(closing chan struct{}) {
+	defer c.wg.Done()
+
+	t := time.NewTicker(c.HeartbeatInterval)
```
Comment (benbjohnson) [inline messaging/client.go]: Add:

```go
defer t.Stop()
```

File: test/messaging.go [hunk 17]
```diff
@@ -9,11 +9,22 @@ import (
 	"github.com/influxdb/influxdb/messaging"
 )
 
+var (
+	testDataURL *url.URL
+)
```
Comment (benbjohnson) [inline test/messaging.go]: Remove `var` block.

Comment (benbjohnson) [inline test/messaging.go]: Actually, why is `testDataURL` a global? Can you add a function `MustParseURL()` and put it wherever it's used? That way people don't have to jump around to understand a test function.

File: test/messaging.go [hunk 18]
```diff
@@ -22,15 +33,21 @@ type MessagingClient struct {
...
 	}
 	c.PublishFunc = c.DefaultPublishFunc
 	c.ConnFunc = c.DefaultConnFunc
 	return c
 }
 
+// NewMessagingClient returns a new instance of MessagingClient.
+func NewTestMessagingClient() *MessagingClient {
+	return NewMessagingClient(*testDataURL)
+}
```
Comment (benbjohnson) [inline test/messaging.go]: Can we change the name of this to `NewDefaultMessagingClient()`? Otherwise it creates a stutter in the caller code: `test.NewTestMessagingClient()` and it's not easy to understand how it's different from the `test.NewMessagingClient()`.
